---
title: "README"
author: "Sylvain Dechaumet"
date: "`r Sys.Date()`"
output: github_document
csl: ./Readme_source/plant-physiology.csl
bibliography: ./Readme_source/rAMOPLS_bibliography.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, message = F, warning = F, fig.path = "Readme_source/figures/", cache = 2)
```


## Overview

The `rAMOPLS` package is a tool for helping you to extract relevant information from multivariate experiments with multiple factors. It provides simple functions to perform and assess an **A**nova **M**ultiblock **O**rthogonal **P**artial **L**east **S**quare analysis **(AMOPLS)** as described by @Boccard2016 with the specific calculation of **V**ariable **I**mportant in the **P**rojection **(VIP²)** described by @Gonzalez-Ruiz2017. Since most biological experiments are unbalanced due to a wide range of reasons, the method of stratified subsampling described by @Boccard2019 has also been implemented here.

Most biologist need to extract relevant information (treatment, tissular, organs, genotype, ...) from noisy and highly uncontrollable experiments. Most of us will try to use a design-of-experiment to assess any factors which can influence the response (wanted and unwanted variations). But then the tools to assess the effect of each factor and their interactions in multivariate experiments are not widely available. The AMOPLS methods combine ANOVA and k-OPLS (@Bylesjo2008) statistical methods to address these problems and provide easy-to-use diagnostics parameters to evaluate the effects.


## Getting started
### Installing

`rAMOPLS` can be installed using one of the following methods:  
  - `remotes`: `remotes::install_github('sdechaumet/rAMOPLS')`  
  - `devtools`: `devtools::install_github('sdechaumet/rAMOPLS')`  
  - `renv`: `renv::install('github:sdechaumet/rAMOPLS')`  
  - `pak`: `pak::pkg_install('sdechaumet/rAMOPLS')`  

The AMOPLS method depends on the kopls package which is available [here](http://kopls.sourceforge.net/index.shtml).
A copy of the package sources is implemented inside the `rAMOPLS` and can be installed with `rAMOPLS::install_kopls()`.

### Calculate an AMOPLS model

The single `run_AMOPLS()` function is a wrapper to perform AMOPLS model with permutations and return loadings, scores, variances, saliences and statistics. It take as input two table (as matrix, data.frame or data.table) with the first column being unique identifier for rows and the column names from samplemetadata to study (factors): 

- `datamatrix`: with observations as rows and measurments as columns  
- `samplemetadata`: with observations as rows and descriptor as columns (sample groups)  
- `factor_names`: Column names of samplemetadata with factor to consider  

Exemple on `data_Ruiz2017` dataset on "Exposure time" and "Dose" with level 1 interaction and 100 permutations: 

```{r}
require(rAMOPLS)
result <- run_AMOPLS(datamatrix = data_Ruiz2017$datamatrix,
                     samplemetadata = data_Ruiz2017$samplemetadata, 
                     factor_names = c("Exposure time", "Dose"))
```


The following arguments are optional *(default value)*: 

- `interaction_level` *(1)*: Order of interactions to consider (0 for main effect and 1 for first order interaction: Fac_A, Fac_B and Fac_A x Fac_B)  
- `scaling` *(T)*: Logical to perform a variance scaling of the datamatrix  
- `nb_perm` *(100)*: Number of permutation to test for each effect (default to 100)  
- `nb_compo_orthos` *(1:3)*: Number of orthogonal component to consider (1:5 to test 1 to 5)  
- `parallel` *(FALSE)*: If TRUE or a number (y), run permutations in parallel one y core using [future](https://cran.r-project.org/web/packages/future/index.html) and [furrr](https://cran.r-project.org/web/packages/furrr/index.html) for progress bar


### Choose the optimal number of orthogonal component

The optimal numnber parsimony principle is applied in the case of equally performing models with a different number of orthogonal components (@Boccard2016): The significative model with the smallest number of orthogonal component is to be choosen.

In this example the first model is already significant and should be choosen.

```{r, out.width="75%", fig.width=8, fig.asp=0.5}
fun_plot_ortho(result)
result_optimal <- result$orthoNb_1
```


### Summary statistics
All graphical output use `ggplot2` and can be customized according to ggplot2 nomenclature.

- Summary statistics:  
```{r, out.width = "100%"}
fun_get_summary(result_optimal) %>% knitr::kable(digits = 3, align = 'c')
```


- Optimal score plot:  
```{r,out.width = "100%", fig.width=12, fig.asp = 0.75}
fun_plot_optimal_scores(result_optimal)
```


- Optimal loading plot:  
```{r, out.width = "100%", fig.width=12, fig.asp = 0.75}
fun_plot_optimal_loadings(result_optimal)
```


- VIP² as described in @Gonzalez-Ruiz2017:  
```{r, out.width = "100%", fig.width=15, fig.asp = 0.75}
fun_plot_VIPs(result_optimal, "Dose")
```


## Stratified subsampling for unbalanced data

Unbalanced experimental designs involve non-orthogonal ANOVA submatrices which can alter effects interpretation (@Drotleff2019). To cope with unbalanced design, the most strategy is to resample the dataset to get a balanced design by either: 

- **oversampling:** complete the group with missing observations by adding existing ones.  
- **undersampling:** delete observations in each group to align with n in the smallest group.  

Since oversampling may alter variance decomposition by incorporating identical samples to the model, @Boccard2019 used a strategy based on undersampling strategy. To cope with minimized dataset (and therfore deleted observation), the subsampling is randomly performed a high number of times (1000s) and the models are combined using ensemble-based estimates from balanced models (involving median calculation of each parameters obtained from AMOPLS).

This strategy as been implemented in rAMOPLS and is automatically triggered if the input dataset are unbalanced. The number of subsampling steps can be customized with the `subsampling` parameter.

Example on the same dataset as @Boccard2019: 

```{r, eval = F}
require(rAMOPLS)
result_unbalanced <- run_AMOPLS(
  datamatrix = data_Boccard2019$datamatrix,
  samplemetadata = data_Boccard2019$samplemetadata, 
  factor_names = c("Chemical", "Batch"), 
  nb_perm = 100,
  subsampling = 10,
  parallel = 3)
result_optimal <- result_unbalanced[[1]]
```

```{r message=FALSE, warning=FALSE, include=FALSE}
require(rAMOPLS)
result_unbalanced <- invisible(
  run_AMOPLS(
    datamatrix = data_Boccard2019$datamatrix,
    samplemetadata = data_Boccard2019$samplemetadata, 
    factor_names = c("Chemical", "Batch"), 
    nb_perm = 100,
    subsampling = 10,
    parallel = 3)
)
result_optimal <- result_unbalanced[[1]]
```

- Summary statistics:  
```{r, out.width = "100%"}
fun_get_summary(result_optimal) %>% knitr::kable(digits = 3, align = 'c')
```


- Optimal score plot:  
```{r,out.width = "100%", fig.width=12, fig.asp = 0.75}
fun_plot_optimal_scores(result_optimal)
```


- Optimal loading plot:  
```{r, out.width = "100%", fig.width=12, fig.asp = 0.75}
fun_plot_optimal_loadings(result_optimal)
```


- VIP² as described in @Gonzalez-Ruiz2017:  
```{r, out.width = "100%", fig.width=15, fig.asp = 0.75}
fun_plot_VIPs(result_optimal, "Chemical")
```



## Model interpretation

The first step is to ensure the consitency of the ANOVA partitionning, the number of orthogonal components to include in the model and the reliability of the observed effects using the following indices: 

- **ANOVA consistency** by using the **R**esidual **S**tructure **R**atio **(RSR)** associated to each effect. This indicator represents the data structure associated to each effect in regard of the residual. Therefore it is an effect-to-residual ratio that reflects the reliability of each effect. Close or equal to 1 mean no effect.  
- **Reliability of the observed effects** is assessed by checking the Realtive Sum of Square (RSS).  
- **Optimal number of orthogonal components** using R²Y and its associated p-value. In the case of equally performing models, the parsimony principle is applied and the least of significative orthogonal component number is choosen. The R²Y indice represents the explained variation of Y An R²Y parameter closer to 1 means a stronger models (more of the predicted variations are captured by the model).  

> Random experimental designs are expected to produce low R2Y values (no model) and RSR indices close to one (no effect). When both the AMOPLS model R2Y and individual RSR indices are deemed significant, biochemical information can be extracted from the loadings of the predictive components (@Boccard2016). 


## To learn more

You can find more detailled explanation on AMOPLS in the references provided at the end of this page, starting by the original article from @Boccard2016.


## Datasets provided

The package is preloaded with the following datasets to test and compare the functionalities with published articles:

- `liver.toxicity`: Microarray of gene expression from Fisher rats after acetaminophen exposition @Bushel2007 and used in @Boccard2016  
- `data_Boccard2016`: Metabolomic profiles of *A. thaliana* leaves from @Boccard2010 used in @Boccard2016  
- `data_Ruiz2017`: Metabolomic profiles of human neural cells from @Gonzalez-Ruiz2017  
- `data_Boccard2019`: Metabolomic profiles used to demonstrate the subsampling strategy in @Boccard2019  


## References
